File: semester_two/algoDat/public/code/vorlesung-07/vorlagen/java/LinkedListTest.java
// Copyright 2017, University of Freiburg,
// Chair of Algorithms and Data Structures.
// Author: Hannah Bast <bast@cs.uni-freiburg.de>,
//         Axel Lehmann <lehmann@cs.uni-freiburg.de>.

import org.junit.Test;
import org.junit.Assert;

/**
 * Tests for our doubly-linked list.
 */
public class LinkedListTest {

  @Test
  public void insertBeforeTest() {
    LinkedListItem itemC = new LinkedListItem("C");  // #1
    LinkedListItem itemA = new LinkedListItem("A");  // #2
    LinkedListItem itemB = new LinkedListItem("B");  // #3
    LinkedListItem itemD = new LinkedListItem("D");  // #4
    LinkedList list = new LinkedList();

    list.insertBefore(itemC, null);
    System.out.println(list.toString());

    list.insertBefore(itemA, itemC);
    System.out.println();
    System.out.println(list.toString());

    list.insertBefore(itemB, itemC);
    System.out.println();
    System.out.println(list.toString());

    list.insertBefore(itemD, null); 
    System.out.println();
    System.out.println(list.toString());

    Assert.assertEquals(list.toString(), "first = #2, last = #4\n"
      + "[this = #2, prev = #0, next = #3, value = \"A\"]\n"
      + "[this = #3, prev = #2, next = #1, value = \"B\"]\n"
      + "[this = #1, prev = #3, next = #4, value = \"C\"]\n"
      + "[this = #4, prev = #1, next = #0, value = \"D\"]");
  }
}

File: semester_two/algoDat/public/code/vorlesung-07/vorlagen/cpp/LinkedListTest.cpp
// Copyright 2017, University of Freiburg,
// Chair of Algorithms and Data Structures.
// Author: Hannah Bast <bast@cs.uni-freiburg.de>.

#include <gtest/gtest.h>
#include <iostream>
#include "./LinkedListItem.h"
#include "./LinkedList.h"

// _____________________________________________________________________________
TEST(LinkedListTest, insertBefore) {
  LinkedListItem itemC("C");  // #1
  LinkedListItem itemA("A");  // #2
  LinkedListItem itemB("B");  // #3
  LinkedListItem itemD("D");  // #4
  LinkedList list;

  list.insertBefore(&itemC, NULL);
  std::cout << list.toString() << std::endl;

  list.insertBefore(&itemA, &itemC);
  std::cout << std::endl;
  std::cout << list.toString() << std::endl;

  list.insertBefore(&itemB, &itemC);
  std::cout << std::endl;
  std::cout << list.toString() << std::endl;

  list.insertBefore(&itemD, NULL);
  std::cout << std::endl;
  std::cout << list.toString() << std::endl;

  ASSERT_STREQ(list.toString().c_str(), "first = #2, last = #4\n"
      "[this = #2, prev = #0, next = #3, value = \"A\"]\n"
      "[this = #3, prev = #2, next = #1, value = \"B\"]\n"
      "[this = #1, prev = #3, next = #4, value = \"C\"]\n"
      "[this = #4, prev = #1, next = #0, value = \"D\"]");
}
